sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -264,23 +264,43 @@ trait NonSQLExpression extends Expression {
   }
 }
 
+/**
+ * Trait for expressions that requires initialization based on the partition index prior to task
+ * execution on a partition.
+ */
+trait PartitionDependent {
```

Comment:
Actually, `Projection` also need the same trait. Shall we call it `PartitionedStateful`?